Add support for i-BUS telemetry

nicole-ashley opened the pull request:
Implements: [#9419: Enhancement Request: iBus/MavLink interoperation & compatiblity](https://github.com/ArduPilot/ardupilot/issues/9419)

Wiki PR: https://github.com/ArduPilot/ardupilot_wiki/pull/6154

Originally based on [#16545: Adding ibus telemetry support via serial to arducopter](https://github.com/ArduPilot/ardupilot/pull/16545) but heavily rewritten and expanded. The old PR has not been updated in two years and I offered to take over the implementation with no objections.

I'm an experienced software engineer but C++ is not my wheelhouse and this is my first contribution to ArduPilot, so I welcome maintainer feedback especially around memory and CPU management, and potential for causing crashes. This code is tested on an fmuv3 board and has so far proven to be stable. Builds based on Stable for SITL and other boards I've seen requested are available here: https://1drv.ms/f/s!AoxGPryyVyPN1pNwsxI6dRJVblKOug?e=YVYKls

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,507 @@
+    uint16_t available = MIN(_port->available(), 1024U);
+    if (available == 0) {
+        return;
+    } else {
```
```suggestion
    }
```
No `else` clauses after a return statement...

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,507 @@
+    if (available == 0) {
+        return;
+    } else {
+        reset_read_state_after_timegap();
```
Please just in-line this.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,507 @@
+        reset_read_state_after_timegap();
+    }
+
+    for (uint16_t i = 0; i < available; i++) {
```
```suggestion
    for (auto i = 0; i < available; i++) {
```

nicole-ashley replied:
I ended up reverting this (well not this one because it's gone but any remaining `for auto` expressions) because it was breaking SITL builds:
```
error: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Werror=sign-compare]
  311 |     for (auto i = 0; i < sizeof(u.buffer) - 2; i++) {

error: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Werror=sign-compare]
  660 |     for (auto i = 0; i < ARRAY_SIZE(mode_map); i++) {
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,507 @@
+    }
+
+    for (uint16_t i = 0; i < available; i++) {
+        const int16_t c = _port->read();
+        if (c < 0) {
+            return;
+        }
```
```suggestion
        uint8_t c;
        if (!_port->read(c)) {
            return;
        }
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,507 @@
+
+    // If we've reached the end of our sensor list, we shouldn't respond; this tells the receiver that there
+    // are no more sensors to discover.
+    if (
+        _incoming_sensor_id > sizeof(_sensors) / sizeof(_sensors[0])
```
```suggestion
    if (_incoming_sensor_id > ARRAY_SIZE(_sensors) ||
```

peterbarker replied:
... actually, change this out for:
```
    const auto *sensor_defintiion = get_sensor_definition(_incoming_sensor_id);
    if (sensor_definition == nullptr || sensor_definition->sensor_type == 0) {
        return;
    }
```

... and change the calls to dependent functions to:
```
 handle_discover_command(*sensor_definition);
```

nicole-ashley replied:
With the change to `static const` that also removes the size ambiguity, meaning this check is now simply `_incoming_sensor_id > ARRAY_SIZE(_sensors)` with no parameter value checks.

However I've retained the `*sensor_definition` variable and passing it to the dependent functions.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,507 @@
+    // are no more sensors to discover.
+    if (
+        _incoming_sensor_id > sizeof(_sensors) / sizeof(_sensors[0])
+        || !get_sensor_definition(_incoming_sensor_id)->sensor_type
```
```suggestion
        || get_sensor_definition(_incoming_sensor_id)->sensor_type == 0
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,507 @@
+
+AP_IBus_Telem::_sensor_definition* AP_IBus_Telem::get_sensor_definition(uint8_t sensor_id)
+{
+    return &_sensors[sensor_id - 1];
```
There will probably only be one caller to this with above-requested changes, and this could probably then be in-lined.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,507 @@
+
+uint16_t AP_IBus_Telem::get_average_cell_voltage_cV()
+{
+    if (AP::battery().has_cell_voltages()) {
```
```suggestion
    if (!AP::battery().has_cell_voltages()) {
        return 0;
    }
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,507 @@
+{
+    if (AP::battery().has_cell_voltages()) {
+        const auto &cell_voltages = AP::battery().get_cell_voltages();
+        const uint8_t number_of_cells = sizeof(cell_voltages.cells) / sizeof(cell_voltages.cells[0]);
```
use ARRAY_SIZE

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,507 @@
+uint16_t AP_IBus_Telem::get_vehicle_mode()
+{
+    const AP_AHRS::VehicleClass vehicle_class = AP::ahrs().get_vehicle_class();
+    if (vehicle_class == AP_AHRS::VehicleClass::GROUND) {
```
`switch`

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.h:
```diff
@@ -0,0 +1,355 @@
+    AP_IBus_Telem() {}
+
+    /* Do not allow copies */
+    AP_IBus_Telem(const AP_IBus_Telem &other) = delete;
+    AP_IBus_Telem &operator=(const AP_IBus_Telem &) = delete;
```
```suggestion
               CLASS_NO_COPY(AP_IBus_Telem);
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.h:
```diff
@@ -0,0 +1,355 @@
+    // i-BUS will generally only query up to 15 sensors, so subjectively
+    // higher-value sensors are sorted to the top to make the most of a
+    // small telemetry window. In the future these could be configurable.
+    _sensor_definition _sensors[25] = {
```
```suggestion
    static const SensorDefinition _sensors[25] = {
```

similarly for the other definitions.

These currently consume RAM, which is bad.  `static const` moves it to just flash.

nicole-ashley replied:
I'm being told I can't initialise these here without also adding `constexpr`; that seems to be working but let me know if there's a preferred way.

nicole-ashley replied:
I notice you changed the case for this struct type, so I've applied that to the other struct types as well. Easy to change it back to snake case if that was accidental.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.h:
```diff
@@ -0,0 +1,355 @@
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Math/vector3.h>
+#include <AP_RPM/AP_RPM.h>
```
```suggestion
#include <AP_RPM/AP_RPM_config.h>
```

peterbarker reviewed (changes requested):
Partial review only, sorry.  This is looking quite reasonable.

Fantastic to get this work!

I have a test suite for IBus which I wrote in Feb 2021 for this which I will get merged (and run this through).

peterbarker commented:
I've fixed up the suite I wrote years ago, and I've PR'd it here: https://github.com/ArduPilot/ardupilot/pull/27695

Needs this patch to run the suite for Copter:
```
--- a/Tools/autotest/arducopter.py
+++ b/Tools/autotest/arducopter.py
@@ -11770,6 +11770,7 @@ class AutoTestCopter(vehicle_test_suite.TestSuite):
             self.InvalidJumpTags,
             self.IMUConsistency,
             self.AHRSTrimLand,
+            self.IBus,
             self.GuidedYawRate,
             self.NoArmWithoutMissionItems,
             self.DO_CHANGE_SPEED_in_guided,
```

And then it's `./Tools/autotest/autotest.py --gdb --debug build.Copter test.Copter.IBus` to run the suite against your code.  Your branch passed that suite when I rebased it on top of my PR branch there.

peterbarker commented:
We'll want an entry in `Tools/scripts/test_build_options.py` and `extract_features.py` to make it available on our custom build server (https://custom.ardupilot.org).  Swipe the lines used for the LTM telemetry backend)

We will probably need to put this into `minimize_common.py` to save the flash on lower-end boards.  Many of these are right on the edge of not compiling - but the custom build server allows users to remove other features to fit things like this in.

nicole-ashley commented:
Thanks @peterbarker for the initial review! I've made all the changes and left you a couple of comments as well. Please let me know if I can contribute to the tests, I realise that a huge amount has changed since that initial PR.

peterbarker commented:
This page links to our standards around how we like our commit history to look: https://ardupilot.org/dev/docs/git-interactive-rebase.html

Basically a commit-per-directory, prefixed with the logical subsystem name.  So you might have `AP_RCProtocol_IBUS: add support for iBus protocol` and `autotest: add tests for iBus porotocol`

nicole-ashley commented:
I've added this to `extract_features.py` and `build_options.py`. Happy to add it to `minimize_common.inc` if you're sure.

nicole-ashley commented:
I've split out the commits. Tools is all in one but I can easily split that further if you'd like.

nicole-ashley commented:
@peterbarker Would you like me to include your tests in this PR as their own commit, or will you keep them separate?

peterbarker commented:
> @peterbarker Would you like me to include your tests in this PR as their own commit, or will you keep them separate?

I intend to merge those tests in that other PR - this PR should include a patch to run them as part of the Copter autotest suite.

Please consider this alternate parser: https://github.com/ArduPilot/ardupilot/commit/391ee28f6714118c2c70f0b80ba6aee97e6ef7ae

We didn't do much frame-gapped parsing way back when.

Advantages of that approach are the lack of a state machine, and not keeping state around in the object which is then assumed by the "handle" methods.

Please note I haven't actually tested that patch I reference as there were link errors on the branch I started from.  I've probably screwed up some endian-ness thing in there somewhere (e.g. the checksumming).  Quite happy to get it fully working if you want to change to that style of parsing rather than the state-machine which is currently in this PR.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,674 @@
+// i-BUS will generally only query up to 15 sensors, so subjectively
+// higher-value sensors are sorted to the top to make the most of a
+// small telemetry window. In the future these could be configurable.
+static const AP_IBus_Telem::SensorDefinition _sensors[] = {
```
```suggestion
static const AP_IBus_Telem::SensorDefinition sensors[] {
```

It doesn't matter here where it's a static-const, but omitting the equals can have benefits - not constructing objects on the stack and increased type-checking.  So just as a pattern we're preferring it.

... also removed the leading underscore on the variable name.  We're moving away from prepending underscores to things - so if we could remove them from the object's variables that would be good...

nicole-ashley replied:
I originally had no underscores but when I did a style guide check I noticed it said to put them in, so I did 😆 

I'll update!

peterbarker replied:
Oh, wow.  Wiki PR coming up for DevCall to fix that, my apologies.

peterbarker replied:
https://github.com/ArduPilot/ardupilot_wiki/pull/6150

nicole-ashley replied:
Wonderful, thank you :)

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,674 @@
+   * 0xFF: Checksum high byte
+   Checksums are 0xFFFF minus the sum of the previous bytes
+ */
+void AP_IBus_Telem::handle_type_command(SensorDefinition sensor)
```
```suggestion
void AP_IBus_Telem::handle_type_command(const SensorDefinition &sensor)
```

& avoids putting data onto the stack, and `const` is just generally useful for keeping logic correct.

similarly elsewhere

nicole-ashley commented:
I much prefer that parser! And it's usually how I would write this kind of thing (I don't like instance-level state). I'll rewrite and test that now (will be tomorrow before I come back with it though).

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,674 @@
+    }
+
+    const auto &cell_voltages = AP::battery().get_cell_voltages();
+    const uint8_t number_of_cells = ARRAY_SIZE(cell_voltages.cells);
```
defensive programming says we should check for zero to avoid the divide-by-zero case below.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,674 @@
+    }
+}
+
+Vector3f AP_IBus_Telem::get_unbiased_acceleration()
+{
+    return (AP::ahrs().get_accel() - AP::ahrs().get_accel_bias());
+}
+
```
Maybe just inline this?  Should save a few bytes

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,674 @@
+    const AP_AHRS::VehicleClass vehicle_class = AP::ahrs().get_vehicle_class();
+
+    switch (vehicle_class) {
+    case AP_AHRS::VehicleClass::GROUND:
```
I *think* we can make this a compile-time decision instead.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,674 @@
+    GUIDED       = 15
+    INITIALISING = 16
+*/
+static const AP_IBus_Telem::ModeMap _ground_vehicle_mode_map[] = {
```
```suggestion
#if APM_BUILD_TYPE(APM_BUILD_ArduPlane)
static const AP_IBus_Telem::ModeMap mode_map[] = {
.
.
};
#elif APM_BUILD_TYPE(APM_BUILD_ArduCopter)
static const AP_IBus_Telem::ModeMap mode_map[] = {
.
.
```

That way only ever have one `mode_map` to look through (avoiding the switch on the AHRS concept of the vehicle)

nicole-ashley replied:
I did originally build it this way but I was getting `@Invalid_use_of_APM_BUILD_TYPE` errors. Maybe that'll have changed now it's in the .cpp file? I'll give it a go.

peterbarker replied:
Yep - the defines I mentioned aren't allowed to be used in headers which aren't on a difficult-to-find whitelist.  It's done this way to try to keep the compilation the same for different vehicles, leading to faster build-times.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.h:
```diff
@@ -0,0 +1,171 @@
+    CLASS_NO_COPY(AP_IBus_Telem);
+
+    void init();
+    void listen();
```
Most of these can be `private`

nicole-ashley commented:
I've made all the changes and they're compiling locally, but I haven't bench-tested them yet. I'll do that tomorrow morning and may have some more tweaks based on my findings.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,681 @@
+        return;
+    }
+
+    const auto *sensor_definition = &sensors[incoming.sensor_id - 1];
```
```suggestion
    const auto &sensor_definition = sensors[incoming.sensor_id - 1];
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,681 @@
+        break;
+
+    case PROTOCOL_COMMAND_TYPE:
+        handle_type_command(incoming, *sensor_definition);
```
```suggestion
        handle_type_command(incoming, sensor_definition);
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,681 @@
+        break;
+
+    case PROTOCOL_COMMAND_VALUE:
+        handle_value_command(incoming, *sensor_definition);
```
```suggestion
        handle_value_command(incoming, sensor_definition);
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.h:
```diff
@@ -0,0 +1,162 @@
+#include <AP_RPM/AP_RPM_config.h>
+
+// 2-byte values
+#define IBUS_SENSOR_TYPE_TEMPERATURE          0x01 // Temperature (in 0.1 degrees, where 0=-40'C)
```
Could move this stuff into the .cpp file - nothing else in ArduPilot needs to know about this.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.h:
```diff
@@ -0,0 +1,162 @@
+
+    void init(); 
+
+    typedef struct {
+        uint8_t sensor_type;   // Sensor type (IBUS_SENSOR_TYPE_* above)
+        uint8_t sensor_length; // Data length for defined sensor (can be 2 or 4 bytes)
+    } SensorDefinition;
+
+    typedef struct {
+        uint8_t ap_mode;
+        uint8_t ibus_mode;
+    } ModeMap;
```
make these private

nicole-ashley replied:
This could be me lacking C++ knowledge here, but if I make them private I'm then told they're inaccessible when using them as types in the .cpp file (as they're being used outside class context, I expect). I could move them out of the class entirely? Or is there a better way?

nicole-ashley replied:
Possibly that's what you mean here? https://github.com/ArduPilot/ardupilot/pull/27695#issuecomment-2258260241

peterbarker replied:
Never mind - it would require additional patches in the cpp file to change

```
   98 | static const AP_IBus_Telem::SensorDefinition sensors[] {
```
to
```
   98 | static const AP_IBus_Telem::SensorDefinition AP_IBus_Telem::sensors[] {
```

and the addition in the header of `sensors`.

The definitions could also be moved to the `cpp` file, of course, but we're starting to get lost in the weeds a bit.

We'll just forget I mentioned this possibility :-)

nicole-ashley replied:
This'll be burning in my mind now so I'm sure I'll come back to it for a future iteration 🙂

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.h:
```diff
@@ -0,0 +1,162 @@
+
+    struct PACKED CommandPacket {
+        uint8_t message_length;
+        uint8_t command : 4;
+        uint8_t sensor_id : 4;
```
```suggestion
        uint8_t sensor_id : 4;
        uint8_t command : 4;
```

... it was close!

peterbarker commented:
Instead of fighting with clang (the cause of the MacOSX compilation failure), move the mode map entirely into the `.cpp` file - don't need namespacing or whatever provided by the header file.

andyp1per commented on libraries/AP_OSD/AP_OSD_ParamSetting.cpp:
```diff
@@ -139,7 +139,7 @@ static const char* SERIAL_PROTOCOL_VALUES[] = {
     "NMEA", "WNDVNE", "SLCAN", "RCIN", "MGSQRT", "LTM", "RUNCAM", "HOT_TLM", "SCRIPT", "CRSF",
     "GEN", "WNCH", "MSP", "DJI", "AIRSPD", "ADSB", "AHRS", "AUDIO", "FETTEC", "TORQ",
-    "AIS", "CD_ESC", "MSP_DP", "MAV_HL", "TRAMP", "DDS", "IMUOUT", "IQ", "PPP",
+    "AIS", "CD_ESC", "MSP_DP", "MAV_HL", "TRAMP", "DDS", "IMUOUT", "IQ", "PPP", "IBUS_TLM"
```
We try and make these 6 chars so that they fit in the various display contexts

nicole-ashley replied:
Would `IBUSTLM` work? There are a few 7char ones. I feel like the "TLM" is important because not all serial protocols are telemetry.

andyp1per replied:
If there are 7 char ones already then ignore me

nicole-ashley replied:
7 is the longest so far. `IBUS_TLM` is 8 which would be longer. I can make it `IBUSTLM` but that breaks convention with the other telemetry ones which all use an underscore. Happy to do whichever you prefer!

andyp1per reviewed (changes requested):
One small comment

nicole-ashley commented:
I've made all the changes other than two I've asked questions on. Happy to make them once I understand them. Thank you for all the C++ and embedded help!

Bench testing on the new parser has passed. It's how I probably would have written it if I started from scratch, I think that was just the last holdover from the original PR that I hadn't rewritten.

I've updated the builds folder with the latest code based on stable-4.5.4.

peterbarker commented:
> I've made all the changes other than two I've asked questions on. Happy to make them once I understand them. Thank you for all the C++ and embedded help!

All good :-)  Really good to get the rapid turn-arounds - that was a lot of churn in a short amount of time, and I know that multiple round-trips with large amounts of requested changes can be frustrating.

> Bench testing on the new parser has passed. It's how I probably would have written it if I started from scratch, I think that was just the last holdover from the original PR that I hadn't rewritten.

:-)

I think @andyp1per had worked on some infrastructure to make writing frame-gapped packet parsing even easier; some sort of an API on the UARTDriver.  I *believe* that one's still a WIP.

> I've updated the builds folder with the latest code based on stable-4.5.4.

OK, this is a bit of a strange comment.  AFAICS this is all based on master.  We only merge features into master, never a stable branch, 'though some get backported after they're in master.  Usually the "builds" folder is a transient directory used by the build system rather than something you'd actively update, too.  So I'm a touch confused :-)

nicole-ashley commented:
And now I can report that about 30 minutes of real-world flight testing has also worked perfectly with no noticable issues or side-effects.

@peterbarker Possibly the build comment is misleading, and it's not important for the purposes of the code review. All the code in this PR is based purely on `master`.

In the opening description I provided a folder with some builds for bench testing. They're this feature based on stable so that any issues noticed are definitely to do with only this feature and not other unreleased changes. This is a holdover from when I was resurrecting the code from the old PR, in case anyone who didn't have a full build environment could help me test it. Possibly there's a more official way to do this but as I didn't have push access to the old PR that was a way to offer up testable binaries. Now that there's been a large functional rewrite (the parser) I updated the builds so that the newest way could easily be tested.

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,730 @@
+        value.uint16 = get_vehicle_mode();
+        break;
+
+    case IBUS_SENSOR_TYPE_SPEED:
+        value.uint16 = AP::airspeed()->get_airspeed();
+        break;
```
```suggestion
#if AP_AIRSPEED_ENABLE
        case IBUS_SENSOR_TYPE_SPEED:
        value.uint16 = AP::airspeed()->get_airspeed();
        break;
#endif
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,730 @@
+        value.uint16 = get_distance_from_home_m();
+        break;
+
+    case IBUS_SENSOR_TYPE_ARMED:
+        value.uint16 = AP::arming().is_armed();
+        break;
```
```suggestion
#if AP_ARMING_ENABLED
    case IBUS_SENSOR_TYPE_ARMED:
        value.uint16 = AP::arming().is_armed();
        break;
#endif
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,730 @@
+
+    case IBUS_SENSOR_TYPE_YAW:
+        value.int16 = AP::ahrs().yaw_sensor;
+        break;
```
```suggestion
        break;
#endif  // AP_AHRS_ENABLED
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,730 @@
+        value.byte[1] = AP::gps().num_sats();
+        break;
+
+    case IBUS_SENSOR_TYPE_ACC_X:
```
```suggestion
#if AP_AHRS_ENABLED
    case IBUS_SENSOR_TYPE_ACC_X:
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,730 @@
+        value.int16 = AP::baro().get_climb_rate() * 100;
+        break;
+
+    case IBUS_SENSOR_TYPE_COG:
```
```suggestion
#if AP_GPS_ENABLED
    case IBUS_SENSOR_TYPE_COG:
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,730 @@
+        value.uint16 = AP::ahrs().yaw_sensor * 0.01;
+        break;
+
+    case IBUS_SENSOR_TYPE_CLIMB_RATE:
```
```suggestion
#if AP_BARO_ENABLED
    case IBUS_SENSOR_TYPE_CLIMB_RATE:
```

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,730 @@
+#if AP_RPM_ENABLED == 1
+    {.sensor_type = IBUS_SENSOR_TYPE_RPM, .sensor_length = 2},
+#endif
+    {.sensor_type = IBUS_SENSOR_TYPE_BATTERY_CURRENT, .sensor_length = 2},
```
```suggestion
#if AP_BATTERY_ENABLED
    {.sensor_type = IBUS_SENSOR_TYPE_BATTERY_CURRENT, .sensor_length = 2},
```

etc

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.h:
```diff
@@ -0,0 +1,113 @@
+
+#pragma once
+
+#ifndef AP_IBUS_TELEM_ENABLED
```
```suggestion
#include <AP_HAL/AP_HAL_Boards.h>

#ifndef AP_IBUS_TELEM_ENABLED
```

will resolve the compile failure with redefinition of this define

peterbarker reviewed (changes requested):
Need to be pretty liberal with cutting out code with the `#if`s.  We have a script to test the build options (`Tools/autotest/test_build_options.py`) which will catch them all, but it takes a long time to run.  I can do that either on the next iteration of this PR or after the thing is merged (i.e. fix them with a follow-up PR.  The concept is that code should compile with every option in `build_options.py` turned off.

A run with `./Tools/autotest/test_build_options.py --no-enable-in-turn --no-disable-in-turn` might save a few round-trips to github trying to get CI to pass.

nicole-ashley commented:
OK! I've made all the changes and also `test_build_options.py --no-enable-in-turn --no-disable-in-turn` has passed locally. So fingers crossed! 🤞

nicole-ashley commented:
The last unaddressed comments are:

@peterbarker I haven't added this to `minimize_common.inc` yet. It'd be sad not to have it in builds automatically but I totally understand why and knowing the build server exists is good. Presumably it will automatically be in builds for boards that aren't so tight on memory? Let me know if you want me to add it to `minimize_common.inc`.

@andyp1per I've left a comment regarding the OSD name; totally happy to take your call.

peterbarker commented:
Closes https://github.com/ArduPilot/ardupilot/issues/9419

tridge commented:
@nicole-ashley very nicely done PR!
My initial reaction is that this should be enabled only via the custom build server, as we are already very low on flash. See https://firmware.ardupilot.org/Tools/BuildSizes/builds.html
What is the typical flight controller used by the community that uses i-Bus? We could enable it on some of the 2M flash boards, but that only makes sense if that overlaps with the likely user community that use i-Bus.
we should also enable it always on SITL so it gets built and doesn't bit rot

peterbarker commented on libraries/AP_IBus_Telem/AP_IBus_Telem.cpp:
```diff
@@ -0,0 +1,796 @@
+    return value;
+}
+
+uint16_t AP_IBus_Telem::get_average_cell_voltage_cV()
+{
+#if AP_BATTERY_ENABLED
```
```suggestion
#if AP_BATTERY_ENABLED
uint16_t AP_IBus_Telem::get_average_cell_voltage_cV()
{
```

... and remove the call

Similarly elsewhere

peterbarker commented:
This is probably the way to limit the boards this is built on: https://github.com/ArduPilot/ardupilot/pull/27717

nicole-ashley commented:
@peterbarker The 2048 suggestion sounds good to me - we get it automatically in the bigger boards and it can still be selected for custom builds.

Thanks @tridge! I'm not actually sure what a common FC for i-BUS users is. I can only speak from my own experience: i-BUS is more commonly in the entry-level TX/RX range, and as I was entry-level myself and didn't have A) lots of money for a FC but also B) enough knowledge to make a call on a very specific or limited design. Therefore I went for something cost effective while still being generic, plug-and-play, and enough memory to know I wouldn't regret it: that lead me to one of the classic Pixhawks running hde fmuv3 build.

The old PR also had interest from users of Pixracer, MatekF405, and SpeedyBeeF405Wing boards.

peterbarker commented:
> @peterbarker I suspect the >=2048 limit will probably rule out most of the boards that are commonly used by owners of i-BUS hardware, meaning they'll need to do a custom build (which is also not an entry-level step). It would be great to get it under 2048 by default. Completely your call though.

Definitely not my call on the 2048 thing - that's tridge as technical lead (the poor bugger).

We did agonize over it a chunk on DevCallEU tonight, musing that it would be great to have a way of working out what to put in the builds which was based on numbers not just gut feel.  We've also mused on integration of the custom build server functionality with MissionPlanner to make the process of adding support for these things more straight-forward.

As tridge mentioned  If we put this on 1MB boards today then we'd instantly break several boards.  If this goes in there by default then something has to go out - and because those things have been in we run the risk of breaking people's working setups, which we don't like doing.

nicole-ashley commented:
@peterbarker @tridge Totally understand with the 1MB boards. It would be great however to see it by default on the 2MB boards (which if I understand correctly would also be excluded by the `>2048` rule). 

However I'm just pleading my case and will absolutely take your (well, @tridge's) call. The important thing is that it's in the core codebase, available for people to use, and even more importantly that it's in the documentation with a reference on how to get it on your board if you want it. Discovery and availability is key even if it's not there for a lot of people by default. I totally get that it is, after all, an edge use case.

nicole-ashley commented:
Of the failing tests:

* sitltest-plane has missing logs but that build is working for me locally (maybe a runner issue?)
* sitltest-copter-tests1d seems to be a timeout possibly from an overloaded runner
* sitltest-copter-tests2b is failing due to the missing i-BUS tests

peterbarker commented:
Please cherry-pick https://github.com/ArduPilot/ardupilot/pull/27717/commits/5bcfd941125470be986b2b28c4d21d49b0634732 to reinstate the feature in SITL / linux / STM32-boards-with-external-flash.

peterbarker reviewed (approved):
LGTM

peterbarker commented:
Merged, thanks!

rmackay9 commented:
Hi @nicole-ashley @peterbarker, @andyp1per,

We've got a user [here in the 4.6 beta testing forums](https://discuss.ardupilot.org/t/copter-4-6-0-beta2-is-available-for-beta-testing/127445/11) strugglingto use this feature.  I wonder if you could help?

Also it looks like we missed adding the new item to the SERIALx_PROTOCOL parameter description

nicole-ashley commented:
@rmackay9 I'm travelling at the moment with intermittent internet access. Hard to tell what's wrong on the face of it but I've suggested some things to check.

I do have my computer so I can make the missing change. A quick pointer about where to do it would be helpful though, or I can spend more time on it when I get home in a few weeks.

rmackay9 commented:
Hi @nicole-ashley, Thanks very much for that!  Hopefully the user will figure it out or maybe we can find someone else to test it and confirm it works or not

nicole-ashley commented:
When I'm back home I can test the new beta as well to make sure everything is still working as expected.
